File: Spring Week 11/examples/longest.cpp
#include<bits/stdc++.h>
using namespace std;

// Input: vector of vector<int> adj; adj[i] stores the neighbors of node i
// Output: vector<int> dist; dist[i] stores the longest path from a 'root' node to i
vector<int> longest(vector<vector<int>> adj){
    int n = adj.size();
    // Counting the number of parents is a quick way to check if
    // removing an ancestor will make the node a root
    vector<int> parent_count(n, 0); 
    queue<int> ready;
    for(int i = 0; i < n; i++){
        for(int j : adj[i]){
            parent_count[j]++;
        }
    }
    for(int i = 0; i < n; i++){
        if(parent_count[i] == 0){ // This is a root
            ready.push(i);
        }
    }
    vector<int> dist(n, 0);
    while(!ready.empty()){
        int node = ready.front();
        ready.pop();
        for(int i : adj[node]){
            dist[i] = max(dist[i], dist[node] + 1);
            parent_count[i]--;
            if(parent_count[i] == 0){ // This is a root now
                ready.push(i);
            }
        }
    }
    return dist;
}

int main(){
    // Input from stdin: n vertices with m edges; will not work if input is not a DAG
    int n,m;
    cin >> n >> m;
    vector<vector<int>> adj(n);
    for(int i = 0; i < m; i++){
        int a,b;
        cin >> a >> b;
        adj[a].push_back(b);
    }
    vector<int> dist = longest(adj);
    for(int i = 0; i < n; i++){
        cout << dist[i] << " ";
    }
    cout << endl;
}

File: Spring Week 11/examples/longest.py
from collections import deque

# Input: list of lists adj; adj[i] stores the neighbors of node i
# Output: list dist; dist[i] stores the longest path from a 'root' node to i
def longest(adj):
    n = len(adj)
    # Counting the number of parents is a quick way to check if
    # removing an ancestor will make the node a root
    parent_count = [0] * n
    ready = deque()
    for i in range(n):
        for j in adj[i]:
            parent_count[j] += 1
    for i in range(n):
        if parent_count[i] == 0: # This is a root
            ready.append(i)

    dist = [0] * n
    while ready:
        node = ready.popleft()
        for i in adj[node]:
            dist[i] = max(dist[i], dist[node] + 1)
            parent_count[i] -= 1
            if parent_count[i] == 0: # This is a root now
                ready.append(i)
    
    return dist


# Input from stdin: n vertices with m edges; will not work if input is not a DAG
n,m = map(int, input().split())
adj = [[] for i in range(n)]
for i in range(m):
   a,b = map(int, input().split())
   adj[a].append(b)
dist = longest(adj)
print(*dist)